xarray/core/dataset.py

```diff
@@ -3291,11 +3298,18 @@ def drop(self, labels, dim=None, *, errors='raise'):
         """
         if errors not in ['raise', 'ignore']:
             raise ValueError('errors must be either "raise" or "ignore"')
-        if utils.is_scalar(labels):
-            labels = [labels]
+
         if dim is None:
+            if isinstance(labels, str) or not isinstance(labels, Iterable):
+                labels = {labels}
+            else:
+                labels = set(labels)
+
             return self._drop_vars(labels, errors=errors)
         else:
+            if utils.is_scalar(labels):
```

Comment:
I don't think I fully understand the use case where dims is not None, where apparently you are allowed to feed into labels a whole DataArray. It also makes me nervous that there's a single unit test, not together with the other tests for Dataset.drop, that triggers the use case.

Reply:
Isn't this for the case you pass a single label; what's the case with passing a whole DataArray?

```python
In [12]: da = xr.DataArray(data=np.ones((2, 3)), dims=['x', 'y'], coords=dict(x=list('ab')))

In [13]: da
Out[13]:
<xarray.DataArray (x: 2, y: 3)>
array([[1., 1., 1.],
       [1., 1., 1.]])
Coordinates:
  * x        (x) <U1 'a' 'b'
Dimensions without coordinates: y

In [14]: da.drop('a', dim='x')
Out[14]:
<xarray.DataArray (x: 1, y: 3)>
array([[1., 1., 1.]])
Coordinates:
  * x        (x) <U1 'b'
Dimensions without coordinates: y
```

Reply:
This is the problem. I'm unsure if it's supposed to be allowed?
I've never seen anywhere else, where dataset variable names are required, the option to pass named DataArray's instead.

https://github.com/pydata/xarray/blob/dba85bdfc1ac311349bbfcabccb02a71747a97b2/xarray/tests/test_dataset.py#L4245

Reply:
The fact is that ``iter(DataArray)`` yields scalar DataArrays, not the elements of the DataArray (unlike ``iter(numpy.ndarray)``)

So the lines
```python
            if isinstance(labels, str) or not isinstance(labels, Iterable):
                labels = {labels}
            else:
                labels = set(labels)
```
fall apart when labels is a DataArray.
To make it work I should change it to something like:
```python
            if isinstance(labels, str) or not isinstance(labels, Iterable):
                labels = {labels}
            elif isinstance(labels, (DataArray, Variable):
                labels = set(labels.values)
            else:
                labels = set(labels)
```

The problem is that the logic above is EVERYWHERE. None of the many, many functions that expect "one or more dimension names" as an input cope (or are tested against) a DataArray input.

I lean towards removing the option entirely, and changing the unit test from
```python
ds.drop(ds.y, dim='y')
```
to
```python
ds.drop(ds.y.values, dim='y')
```

This is chiefly due to the fact that passing a DataArray is incredibly counter-intuitive; it took me a *long* time to figure out that ``ds.drop(ds.y, dim='y')`` is not trying to drop the y variable.

Reply:
Note that I wrote ``labels = labels.values`` and not ``labels = set(labels.values)``. That would cause a huge performance hit when values has many elements.

Reply:
I think the problem is due to the fact that the same method drop() performs two fundamentally different things - drop variables or elements from the variables - and in the latter case there are plenty of similar methods that DO accept a DataArray, e.g. sel(), where(), etc.

Reply:
I think I cracked it - please review

Reply:
Yes, it would be ideal to have separate functions for dropping variables rather than elements along an array dimension.

We are also considering refactoring this function over in https://github.com/pydata/xarray/pull/3128, I would be very interested in your thoughts!